src/shogun/lib/SGMatrix.h

```diff
@@ -309,6 +333,68 @@ template<class T> class SGMatrix : public SGReferencedData
 		/** fill matrix with zeros */
 		void zero();
 
+		/**
+		 * fill matrix with uniformly distributed randoms
+		 * Float Types: [0, 1)
+		 * Int Types: std::numeric_limit<T>::(min->max)
+		 * Rest: disabled
+		**/
+		// vectorized available
+		template <typename U>
+		using enable_simd_float = std::enable_if_t<is_any<U, float64_t>::value>;
+		template <typename U = T>
+		auto random() -> enable_simd_float<U>
```

Comment:
separating declaration and definition would need explicit specialization of each template method signature - which would end up being really long. Since the methods are relatively short it seemed ok to keep both in the header file.

The only disadvantage is that an additional include of Math.h has to be added here

Reply:
Yep - Actually 3 versions are possible which mean the same:
1. https://coliru.stacked-crooked.com/a/bc08035924d3582e
2. https://coliru.stacked-crooked.com/a/7cc61a4adcdd5ded
3. https://coliru.stacked-crooked.com/a/45dba741ddf2d4a0

Mine is a slightly more readable version of (2) :D

Reply:
OK! It is more readable indeed! Why do you need the `template <typename U = T>` though? Doesn't `template <typename T>` work?

Reply:
SFINAE only works on templated functions. A method inside a template class doesnt count. The function itself has to also be templated. As mentioned in [CPPReference](https://en.cppreference.com/w/cpp/language/sfinae) `This rule applies during overload resolution of function templates`

Reply:
Btw I wrote up a short blog post on this topic - more as a point of reference but [here it is](https://saatvikshah1994.github.io/TemplateOverloadResolution/) in case anyone is interested :)

Reply:
Ah true! Thanks, I'll have a look!

Reply:
Ah cool!